pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/TemplateDao.java

```diff
@@ -81,57 +84,97 @@ default Long insertTemplateIfNotNull(Template template, long revisionId) {
+    default Long[] insertTemplates(List<Template> templates, long revisionId) {
+        // it is helpful to accept some nulls in the template list, but lets filter those out
+        List<Template> nonNullTemplates = templates.stream().filter(Objects::nonNull).collect(toList());
+        if (nonNullTemplates.stream().anyMatch(t -> t.getTemplateId() != null)) {
+            throw new IllegalStateException("One of the templates had a Template id");
         }
+        for (Template template : nonNullTemplates) {
+            for (TemplateVariable variable : template.getVariables()) {
+                String variableName = variable.getName();
 
+                if (I18nTemplateConstants.DDP.equals(variableName) || I18nTemplateConstants.LAST_UPDATED.equals(variableName)) {
+                    throw new DaoException("Variable name '" + variableName + "' is not allowed");
+                }
+            }
+        }
         JdbiTemplate jdbiTemplate = getJdbiTemplate();
         JdbiTemplateVariable jdbiTemplateVariable = getJdbiTemplateVariable();
         JdbiVariableSubstitution jdbiVariableSubstitution = getJdbiVariableSubstitution();
 
-        TemplateType templateType = template.getTemplateType();
-        String templateText = template.getTemplateText();
-        String templateCode = template.getTemplateCode();
+        // Note we are just generating the guids without checking the database. For our purpose the very rare collision
+        // is a tolerable problem
+        List<String> templateCodes = nonNullTemplates.stream().map(t -> DBUtils.generateUncheckedStandardGuid()).collect(toList());
+        Iterator<TemplateType> templateTypes = nonNullTemplates.stream().map(Template::getTemplateType).iterator();
+        Iterator<String> templateText = nonNullTemplates.stream().map(Template::getTemplateText).iterator();
 
-        if (templateCode == null) {
-            templateCode = jdbiTemplate.generateUniqueCode();
-            template.setTemplateCode(templateCode);
+        long[] templateIds = jdbiTemplate.insert(templateCodes, templateTypes, templateText, revisionId);
+
+        for (int i = 0; i < templateIds.length; i++) {
+            nonNullTemplates.get(i).setTemplateId(templateIds[i]);
+            nonNullTemplates.get(i).setTemplateCode(templateCodes.get(i));
         }
+        class IdToTemplateVariable {
+            Long templateId;
+            TemplateVariable variable;
 
-        long templateId = jdbiTemplate.insert(templateCode,
-                                                templateType,
-                                                templateText,
-                                                revisionId);
-        template.setTemplateId(templateId);
+            IdToTemplateVariable(Long id, TemplateVariable variable) {
+                this.templateId = id;
+                this.variable = variable;
+            }
+        }
 
```

Comment:
Tried to use the Tuples library to keep pairs of ids and variables, but turned difficult to use with the lambda type checking. Just created my own local class for purpose.